subprojects/core/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationRolesForMigration.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.configurations;
+
+/**
+ * The roles here all meant to be temporary roles used for migration only, to be removed in Gradle 9.0.
+ * <p>
+ * While we currently (Gradle 8.x) have to support the legacy behavior of configurations, we want to encode
+ * the knowledge of what the minimally required intended behavior is, so that we easily migrate to the
+ * {@link #eventualRole} in Gradle 9.0.
+ */
+public enum ConfigurationRolesForMigration implements ConfigurationRole {
+    @Deprecated
+    LEGACY_TO_INTENDED_RESOLVABLE_BUCKET(ConfigurationRoles.LEGACY, ConfigurationRoles.INTENDED_RESOLVABLE_BUCKET),
+    @Deprecated
+    LEGACY_TO_INTENDED_CONSUMABLE(ConfigurationRoles.LEGACY, ConfigurationRoles.INTENDED_CONSUMABLE),
+
+    @Deprecated
+    INTENDED_RESOLVABLE_BUCKET_TO_INTENDED_RESOLVABLE(ConfigurationRoles.INTENDED_RESOLVABLE_BUCKET, ConfigurationRoles.INTENDED_RESOLVABLE),
```

Comment:
We wanted to leave the `INTENDED` prefix in the `ConfigurationRoles`, and I want these names to exactly match those, to avoid any chance of confusion.

I see the long, ugly names, and the deprecation as features - they make it very obvious that this is something temporary and not to be imitated.  I'd rather leave the names and deprecation unwieldy, and focus on removing them as soon as we can.